File: v0.2/golang/scenario/flow.go
// SPDX-License-Identifier: Apache-2.0
//
// Based on the Provenance Identity Continuity (PIC) Model created by Nicola Gallo.
// Conforms to the PIC Specification published and maintained by Nitro Agility S.r.l.

package scenario

import (
	"time"

	"github.com/pic-protocol/pic-prototyping/v0.2/golang/pic"
)

// This file builds one end-to-end execution flow: an origin PCA0 whose authority
// narrows hop by hop as real fixture executors continue it, plus a final rogue
// attempt that PIC rejects. It backs the `picdemo flow` visualization.

// FlowHop is one step of the execution flow: who acted, what they did, the
// authority they carry, what they dropped versus their predecessor, and the real
// signed PCA they produced.
type FlowHop struct {
	Index          int      `json:"hop"`
	Actor          string   `json:"actor"`
	Action         string   `json:"action"`
	Authority      []string `json:"authority"`
	Dropped        []string `json:"dropped,omitempty"`
	LineageCounter uint64   `json:"lineageCounter"`
	PreviousHash   string   `json:"previousPcaHash,omitempty"`
	Generates      *pic.PCA `json:"generates"`
}

// RogueAttempt is a compromised executor trying to re-expand authority; PIC
// rejects it (non-expansion), which is the point.
type RogueAttempt struct {
	Actor    string   `json:"actor"`
	Tried    []string `json:"tried"`
	Rejected bool     `json:"rejected"`
	Reason   string   `json:"reason"`
}

// FlowResult is the whole flow: the hops, the verification outcome, and the
// rejected rogue attempt.
type FlowResult struct {
	Description  string        `json:"description"`
	Hops         []FlowHop     `json:"hops"`
	VerifyOK     bool          `json:"verifyOk"`
	TipAuthority []string      `json:"tipAuthority"`
	Rogue        *RogueAttempt `json:"rogue"`
}

// Flow runs the execution flow on the real fixtures and returns it. Every PCA is
// really minted, signed, and verified on this call.
func (w *World) Flow(now time.Time) (*FlowResult, error) {
	contract := pic.ExecutionContract{} // permissive, so every fixture executor conforms
	ops0 := []string{"read-all", "backup", "share-files"}

	pca0, err := pic.MintPCA0(w.id("alice"), pic.Invariants{Operations: ops0, ExecutionContract: contract}, "", now)
	if err != nil {
		return nil, err
	}
	chain := []*pic.PCA{pca0}

	res := &FlowResult{
		Description: "Authority created once at the origin (alice) and propagated, only narrowing, through a causal chain of real fixture executors. Each hop produces a signed PCA; the final rogue expansion is rejected.",
		Hops: []FlowHop{{
			Index: 0, Actor: "alice", Action: "mint origin PCA0",
			Authority: ops0, LineageCounter: 0, Generates: pca0,
		}},
	}

	steps := []struct {
		actor, action, op string
		ops               []string
	}{
		{"gateway", "continue: forward (no attenuation)", "read-all", []string{"read-all", "backup", "share-files"}},
		{"backup-service", "continue: attenuate (drop share-files)", "backup", []string{"read-all", "backup"}},
		{"archive-service", "continue: attenuate (drop backup)", "read-all", []string{"read-all"}},
		{"storage-service", "execute: read under {read-all}", "read-all", []string{"read-all"}},
	}
	for i, s := range steps {
		pred := chain[len(chain)-1]
		req := pic.Request{Operation: s.op, Target: "eu-1/tenant-42/resource", SecurityDomain: "tenant-42"}
		next, err := pic.NewProver(w.id(s.actor), w.att(s.actor)).Continue(pred, pic.Invariants{Operations: s.ops, ExecutionContract: contract}, req, now)
		if err != nil {
			return nil, err
		}
		chain = append(chain, next)
		ph, _ := pred.Digest()
		res.Hops = append(res.Hops, FlowHop{
			Index: i + 1, Actor: s.actor, Action: s.action,
			Authority: s.ops, Dropped: diffOps(pred.Invariants.Operations, s.ops),
			LineageCounter: next.LineageCounter, PreviousHash: ph, Generates: next,
		})
	}

	inv, verr := w.verifier().VerifyFullChain(chain, now)
	res.VerifyOK = verr == nil
	res.TipAuthority = inv.Operations

	// Rogue: a compromised executor tries to re-add 'backup' the lineage dropped.
	tip := chain[len(chain)-1]
	tried := []string{"read-all", "backup"}
	rogue, err := pic.NewProver(w.id("archive-service"), w.att("archive-service")).
		ContinueMalicious(tip, pic.Invariants{Operations: tried, ExecutionContract: contract},
			pic.Request{Operation: "backup", Target: "eu-1/tenant-42/resource", SecurityDomain: "tenant-42"}, now)
	if err != nil {
		return nil, err
	}
	rerr := w.verifier().VerifyHop(rogue, tip, now, false)
	res.Rogue = &RogueAttempt{Actor: "archive-service", Tried: tried, Rejected: rerr != nil, Reason: errStr(rerr)}
	return res, nil
}

// diffOps returns the operations present in a but not in b.
func diffOps(a, b []string) []string {
	var out []string
	for _, x := range a {
		found := false
		for _, y := range b {
			if x == y {
				found = true
				break
			}
		}
		if !found {
			out = append(out, x)
		}
	}
	return out
}

func errStr(err error) string {
	if err == nil {
		return ""
	}
	return err.Error()
}

File: v0.2/rust/src/scenario/flow.rs
// SPDX-License-Identifier: Apache-2.0
//
// Based on the Provenance Identity Continuity (PIC) Model created by Nicola Gallo.
// Conforms to the PIC Specification published and maintained by Nitro Agility S.r.l.

//! One end-to-end execution flow: an origin PCA0 whose authority narrows hop by
//! hop as real fixture executors continue it, plus a final rogue attempt that PIC
//! rejects. Backs the `picdemo flow` visualization.

use super::World;
use crate::prover::mint_pca0;
use crate::types::{ExecutionContract, Invariants, Pca, Request};
use crate::{PicResult, Prover};
use chrono::{DateTime, Utc};
use serde::Serialize;

/// One step of the execution flow: who acted, what they did, the authority they
/// carry, what they dropped versus their predecessor, and the signed PCA.
#[derive(Serialize)]
pub struct FlowHop {
    #[serde(rename = "hop")]
    pub index: usize,
    pub actor: String,
    pub action: String,
    pub authority: Vec<String>,
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub dropped: Vec<String>,
    #[serde(rename = "lineageCounter")]
    pub lineage_counter: u64,
    #[serde(rename = "previousPcaHash", skip_serializing_if = "String::is_empty")]
    pub previous_hash: String,
    pub generates: Pca,
}

/// A compromised executor trying to re-expand authority; PIC rejects it.
#[derive(Serialize)]
pub struct RogueAttempt {
    pub actor: String,
    pub tried: Vec<String>,
    pub rejected: bool,
    pub reason: String,
}

/// The whole flow: the hops, the verification outcome, and the rejected rogue.
#[derive(Serialize)]
pub struct FlowResult {
    pub description: String,
    pub hops: Vec<FlowHop>,
    #[serde(rename = "verifyOk")]
    pub verify_ok: bool,
    #[serde(rename = "tipAuthority")]
    pub tip_authority: Vec<String>,
    pub rogue: Option<RogueAttempt>,
}

struct Step {
    actor: &'static str,
    action: &'static str,
    op: &'static str,
    ops: Vec<String>,
}

impl World {
    /// Runs the execution flow on the real fixtures. Every PCA is really minted,
    /// signed, and verified on this call.
    pub fn flow(&self, now: DateTime<Utc>) -> PicResult<FlowResult> {
        let contract = ExecutionContract::default(); // permissive; every executor conforms
        let ops0 = vec![
            "read-all".to_string(),
            "backup".to_string(),
            "share-files".to_string(),
        ];

        let pca0 = mint_pca0(
            self.id("alice"),
            Invariants {
                operations: ops0.clone(),
                execution_contract: contract.clone(),
            },
            "",
            now,
        );
        let mut chain = vec![pca0.clone()];

        let mut res = FlowResult {
            description: "Authority created once at the origin (alice) and propagated, only narrowing, through a causal chain of real fixture executors. Each hop produces a signed PCA; the final rogue expansion is rejected.".to_string(),
            hops: vec![FlowHop {
                index: 0,
                actor: "alice".to_string(),
                action: "mint origin PCA0".to_string(),
                authority: ops0,
                dropped: Vec::new(),
                lineage_counter: 0,
                previous_hash: String::new(),
                generates: pca0,
            }],
            verify_ok: false,
            tip_authority: Vec::new(),
            rogue: None,
        };

        let steps = [
            Step {
                actor: "gateway",
                action: "continue: forward (no attenuation)",
                op: "read-all",
                ops: vec![
                    "read-all".to_string(),
                    "backup".to_string(),
                    "share-files".to_string(),
                ],
            },
            Step {
                actor: "backup-service",
                action: "continue: attenuate (drop share-files)",
                op: "backup",
                ops: vec!["read-all".to_string(), "backup".to_string()],
            },
            Step {
                actor: "archive-service",
                action: "continue: attenuate (drop backup)",
                op: "read-all",
                ops: vec!["read-all".to_string()],
            },
            Step {
                actor: "storage-service",
                action: "execute: read under {read-all}",
                op: "read-all",
                ops: vec!["read-all".to_string()],
            },
        ];

        for (i, s) in steps.into_iter().enumerate() {
            let pred = chain.last().unwrap().clone();
            let req = Request {
                operation: s.op.to_string(),
                target: "eu-1/tenant-42/resource".to_string(),
                security_domain: "tenant-42".to_string(),
                ..Default::default()
            };
            let next = Prover::new(self.id(s.actor), self.att(s.actor)).continue_(
                &pred,
                Invariants {
                    operations: s.ops.clone(),
                    execution_contract: contract.clone(),
                },
                req,
                now,
            )?;
            let ph = pred.digest();
            let dropped = diff_ops(&pred.invariants.operations, &s.ops);
            res.hops.push(FlowHop {
                index: i + 1,
                actor: s.actor.to_string(),
                action: s.action.to_string(),
                authority: s.ops.clone(),
                dropped,
                lineage_counter: next.lineage_counter,
                previous_hash: ph,
                generates: next.clone(),
            });
            chain.push(next);
        }

        let mut v = self.verifier();
        match v.verify_full_chain(&chain, now) {
            Ok(inv) => {
                res.verify_ok = true;
                res.tip_authority = inv.operations;
            }
            Err(_) => {
                res.verify_ok = false;
                res.tip_authority = Vec::new();
            }
        }

        // Rogue: a compromised executor tries to re-add 'backup' the lineage dropped.
        let tip = chain.last().unwrap().clone();
        let tried = vec!["read-all".to_string(), "backup".to_string()];
        let rogue = Prover::new(self.id("archive-service"), self.att("archive-service"))
            .continue_malicious(
                &tip,
                Invariants {
                    operations: tried.clone(),
                    execution_contract: contract.clone(),
                },
                Request {
                    operation: "backup".to_string(),
                    target: "eu-1/tenant-42/resource".to_string(),
                    security_domain: "tenant-42".to_string(),
                    ..Default::default()
                },
                now,
            )?;
        let rerr = self.verifier().verify_hop(&rogue, &tip, now, false).err();
        res.rogue = Some(RogueAttempt {
            actor: "archive-service".to_string(),
            tried,
            rejected: rerr.is_some(),
            reason: rerr.unwrap_or_default(),
        });
        Ok(res)
    }
}

/// Returns the operations present in `a` but not in `b`.
fn diff_ops(a: &[String], b: &[String]) -> Vec<String> {
    a.iter().filter(|x| !b.contains(x)).cloned().collect()
}
